src/crafting.h

```diff
@@ -104,6 +104,84 @@ struct recipe {
 
 };
 
+/**
+*   enum used by comp_selection to indicate where a component should be consumed from.
+*/
+enum usage {
+    use_from_map = 1,
+    use_from_player = 2,
+    use_from_both = 1 | 2,
+    use_from_none = 4,
+    cancel = 5 // FIXME: hacky.
+};
+
+/**
+*   Struct that represents a selection of a component for crafting.
+*/
+template<typename CompType = component>
+struct comp_selection {
+    /** Tells us where the selected component should be used from. */
+    usage use_from = use_from_none;
+    CompType comp;
+
+    /** provides a translated name for 'comp', suffixed with it's location e.g '(nearby)'. */
+    std::string nname() const;
+};
+using item_selection = comp_selection<item_comp>;
+using tool_selection = comp_selection<tool_comp>;
+
+/**
+*   Class that describes a crafting job.
+*
+*   The class has functions to execute the crafting job.
+*/
+class craft_command {
```

Comment:
This is a super good idea, it makes it a lot easier to add features in the future like having NPCs take over crafting jobs for you or vice versa.